File: poo/python/Payroll/HourlyEmployee.py
#!/usr/bin/env python
#-*- coding: utf-8 -*-

"""
@author Nathan Cruz <nathan.cruz@unifesp.br>
@created 14/04/14
"""
from __future__ import unicode_literals
from __future__ import print_function

from datetime import date
from Employee import Employee

class HourlyEmployee(Employee):

    # Atributos #
    __wage = None
    __hours = None

    # Construtor #
    def __init__(self, first, last, ssn, birthDate, departmentCode,
                 hourlyWage, hoursWorked):
        super(HourlyEmployee, self).__init__(first, last, ssn, birthDate,
                                             departmentCode)
        self.setWage(hourlyWage)
        self.setHours(hoursWorked)

    # Getters #
    def getWage(self):
        return self.__wage
    def getHours(self):
        return self.__hours

    # Setters #
    def setWage(self, wage):
        if wage < 0:
            raise ValueError('Cannot set wage to negative value')
        self.__wage = wage

    def setHours(self, hours):
        if hours < 0:
            raise ValueError('Cannor set hours to negative value')
        self.__hours = hours

    def earnings(self):
        hj = date.today()
        pai = super(HourlyEmployee, self)
        bonus = 0
        if hj.month == pai.getPrivateBirthDate().month:
            bonus = 150
        if self.getHours() <= 40:
            return self.getWage() * self.getHours + bonus
        else:
            return (40* self.getWage() + 
                    (self.getHours - 40) * self.getWage() * 1.5) + bonus

    def __repr__(self):
        return self.__str__()

    def __str__(self):
        pai = super(HourlyEmployee, self)
        return '''hourly employee: {0}
                  {1}: ${2}; {3}: {4}'''.format(str(pai), 'hourly wage',
                                                self.getWage(), 'hours worked'
                                                , self.getHours())

File: poo/java/Payroll/HourlyEmployee.java
import java.util.Date;

public class HourlyEmployee extends Employee{

    /* Atributos */
    private double wage;
    private double hours;

    /* Construtor */
    public HourlyEmployee(String first, String last, String ssn,
                         Date birthDate, int departmentCode,
                         double hourlyWage, double hoursWorked){
        super(first, last, ssn, birthDate, departmentCode);
        this.setWage(hourlyWage);
        this.setHours(hoursWorked);
    }

    /* Getters */
    public double getWage(){
        return wage;
    } 
    public double getHours(){
        return hours;
    }

    /* Setters */
    public void setWage(double hourlyWage){
        wage = ( hourlyWage < 0.0 ) ? 0.0 : hourlyWage;
    }
    public void setHours( double hoursWorked )
    {
        this.hours = ((hoursWorked >= 0.0) && 
                      ( hoursWorked <= 168.0)) ? hoursWorked : 0.0;
    }

    /* Calcula os rendimentos do empregado */
    public double earnings(){
        Date today=  new Date();
        float bonus = 0;
        if(today.getMonth() == super.getPrivateBirthDate().getMonth())
            bonus = 150;
        if (getHours() <= 40) return getWage() * getHours() + bonus;
        else 
            return (40 * getWage() + ( getHours() - 40 ) * getWage() * 1.5) + 150;
    }              

    /* Metodo invocado quando tentamos imprimir objetos desta classe */
    public String toString(){
        return String.format("hourly employee: %s\n%s: $%,.2f; %s: %,.2f",
                             super.toString(), "hourly wage", getWage(),
                             "hours worked", getHours());
    }
}
